File: 1101-1200/1101-The Earliest Moment When Everyone Become Friends/1101-The Earliest Moment When Everyone Become Friends.py
class Solution:
    def earliestAcq(self, logs: List[List[int]], N: int) -> int:
        parent = [i for i in range(N)]
        def findParent(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        count = N
        for time, a, b in sorted(logs):
            pa = findParent(a)
            pb = findParent(b)
            if pa != pb:
                count -= 1
                if count == 1:
                    return time
                parent[pb] = pa
        return -1

File: 1101-1200/1101-The Earliest Moment When Everyone Become Friends/1101-The Earliest Moment When Everyone Become Friends.cpp
class Solution {
public:
    int earliestAcq(vector<vector<int>>& logs, int N) {
        vector<int> parent(N);
        for (int i = 0; i < N; ++i) {
            parent[i] = i;
        }
        
        sort(logs.begin(), logs.end());
        int count = N;
        for (auto log : logs) {
            int pa = findParent(parent, log[1]);
            int pb = findParent(parent, log[2]);
            if (pa != pb) {
                if (--count == 1) {
                    return log[0];
                }
                parent[pb] = pa;
            }
        }
        
        return -1;
    }
    
private:
    int findParent(vector<int>& parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
};
